pandas/core/interval.py

```diff
@@ -0,0 +1,521 @@
+        name : object, optional
+            Name to be stored in the index.
+
+        Examples
+        --------
+
+        >>> IntervalIndex.from_breaks([0, 1, 2, 3])
+        IntervalIndex(left=[0, 1, 2],
+                      right=[1, 2, 3],
+                      closed='right')
+        """
+        return cls(breaks[:-1], breaks[1:], closed, name)
+
+    @classmethod
+    def from_intervals(cls, data, name=None):
+        """
+        Construct an IntervalIndex from a 1d array of Interval objects
+
+        Parameters
+        ----------
+        data : array-like (1-dimensional)
+            Array of Interval objects. All intervals must be closed on the same
+            sides.
+        name : object, optional
+            Name to be stored in the index.
+
+        Examples
+        --------
+
+        >>> IntervalIndex.from_intervals([Interval(0, 1), Interval(1, 2)])
+        IntervalIndex(left=[0, 1],
+                      right=[1, 2],
+                      closed='right')
+
+        The generic Index constructor work identically when it infers an array
+        of all intervals:
+
+        >>> Index([Interval(0, 1), Interval(1, 2)])
+        IntervalIndex(left=[0, 1],
+                      right=[1, 2],
+                      closed='right')
+        """
+        data = np.asarray(data)
+        left, right, closed = intervals_to_interval_bounds(data)
+        return cls(left, right, closed, name)
+
+    @classmethod
+    def from_tuples(cls, data, closed='right', name=None):
+        left = []
+        right = []
+        for l, r in data:
+            left.append(l)
+            right.append(r)
+        return cls(np.array(left), np.array(right), closed, name)
+
+    def to_tuples(self):
+        return Index(com._asarray_tuplesafe(zip(self.left, self.right)))
+
+    @cache_readonly
+    def _multiindex(self):
+        return MultiIndex.from_arrays([self.left, self.right],
+                                      names=['left', 'right'])
+
+    @property
+    def left(self):
+        return self._left
+
+    @property
+    def right(self):
+        return self._right
+
+    @property
+    def closed(self):
+        return self._closed
+
+    def __len__(self):
+        return len(self.left)
+
+    @cache_readonly
+    def values(self):
+        """Returns the IntervalIndex's data as a numpy array of Interval
+        objects (with dtype='object')
+        """
+        left = np.asarray(self.left)
+        right = np.asarray(self.right)
+        return interval_bounds_to_intervals(left, right, self.closed)
+
+    def __array__(self, result=None):
+        """ the array interface, return my values """
+        return self.values
+
+    def __array_wrap__(self, result, context=None):
+        # we don't want the superclass implementation
+        return result
+
+    def _array_values(self):
+        return self.values
+
+    def __reduce__(self):
+        return self.__class__, (self.left, self.right, self.closed, self.name)
```

Comment:
I would use a dict-like to capture attributes, this is from ``DatetimeIndex``
```
    def __reduce__(self):

        # we use a special reudce here because we need
        # to simply set the .tz (and not reinterpret it)

        d = dict(data=self._data)
        d.update(self._get_attributes_dict())
        return _new_DatetimeIndex, (self.__class__, d), None
```